File: Mandatories/_Mandatory_I.md/01a/Node.js/parseFiles.js
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const xml2js = require('xml2js');
const csv = require('csv-parser');

// Define the correct path to the data folder
const dataDir = path.join(__dirname, '../data');

// Function to read and parse a text file
const parseText = (filePath) => {
    return fs.readFileSync(filePath, 'utf-8').split('\n').map(line => line.trim()).filter(line => line);
};

// Function to read and parse a JSON file
const parseJSON = (filePath) => {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
};

// Function to read and parse a YAML file
const parseYAML = (filePath) => {
    return yaml.load(fs.readFileSync(filePath, 'utf-8'));
};

// Function to read and parse a CSV file
const parseCSV = (filePath) => {
    return new Promise((resolve) => {
        let results = [];
        fs.createReadStream(filePath)
            .pipe(csv())
            .on('data', (data) => results.push(data))
            .on('end', () => resolve(results));
    });
};

// Function to read and parse an XML file
const parseXML = async (filePath) => {
    const xmlData = fs.readFileSync(filePath, 'utf-8');
    const parser = new xml2js.Parser();
    return parser.parseStringPromise(xmlData).then(result => result.fruits.fruit);
};

// File paths using the correct data directory
const files = {
    text: path.join(dataDir, "sample.txt"),
    json: path.join(dataDir, "sample.json"),
    yaml: path.join(dataDir, "sample.yaml"),
    csv: path.join(dataDir, "sample.csv"),
    xml: path.join(dataDir, "sample.xml")
};

// Parsing and printing results
(async () => {
    console.log("Text:", parseText(files.text));
    console.log("JSON:", parseJSON(files.json));
    console.log("YAML:", parseYAML(files.yaml));
    console.log("CSV:", await parseCSV(files.csv));
    console.log("XML:", await parseXML(files.xml));
})();

File: Mandatories/_Mandatory_I.md/01a/Python/parse_files.py
import json
import yaml
import csv
import xml.etree.ElementTree as ET

# Read and parse a text file
def parse_text(file_path):
    with open(file_path, 'r') as f:
        content = f.read()
    return content

# Read and parse an XML file
def parse_xml(file_path):
    tree = ET.parse(file_path)
    root = tree.getroot()
    data = [{child.tag: child.text for child in item} for item in root.findall("item")]
    return data

# Read and parse a YAML file
def parse_yaml(file_path):
    with open(file_path, 'r') as f:
        data = yaml.safe_load(f)
    return data

# Read and parse a JSON file
def parse_json(file_path):
    with open(file_path, 'r') as f:
        data = json.load(f)
    return data

# Read and parse a CSV file
def parse_csv(file_path):
    with open(file_path, newline='') as f:
        reader = csv.DictReader(f)
        data = [row for row in reader]
    return data

# File paths (adjust accordingly)
files = {
    "Text": "sample.txt",
    "XML": "sample.xml",
    "YAML": "sample.yaml",
    "JSON": "sample.json",
    "CSV": "sample.csv"
}

# Parsing and printing results
print("Text File:", parse_text(files["Text"]))
print("XML File:", parse_xml(files["XML"]))
print("YAML File:", parse_yaml(files["YAML"]))
print("JSON File:", parse_json(files["JSON"]))
print("CSV File:", parse_csv(files["CSV"]))
